File: android/app/src/main/java/com/pabl0sk1/jarvis/Memoria.kt
package com.pabl0sk1.jarvis

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.time.LocalDate

/**
 * Memoria a largo plazo: lo que Jarvis sabe del usuario entre una sesión y otra.
 *
 * Además de su archivo privado, guarda una copia en Android/data/com.pabl0sk1.jarvis/files,
 * que se puede copiar al PC con respaldar_memoria.bat. Si la app se reinstala y esa copia
 * se vuelve a poner en su sitio (restaurar_memoria.bat), la importa al arrancar.
 */
class Memoria(contexto: Context) {

    private val archivo = File(contexto.filesDir, NOMBRE)
    private val copia: File? = contexto.getExternalFilesDir(null)?.let { File(it, NOMBRE) }
    private val datos = mutableListOf<Pair<String, String>>()  // (dato, fecha)

    init {
        val origen = when {
            archivo.exists() -> archivo
            copia?.exists() == true -> copia  // restaurada desde el PC tras reinstalar
            else -> null
        }
        if (origen != null) {
            val lista = JSONArray(origen.readText())
            for (i in 0 until lista.length()) {
                val d = lista.getJSONObject(i)
                datos += d.getString("dato") to d.optString("fecha")
            }
            if (origen != archivo || copia?.exists() != true) guardar()
        }
    }

    @Synchronized
    fun recordar(dato: String) {
        datos += dato.trim() to LocalDate.now().toString()
        guardar()
    }

    /** Borra los recuerdos que contienen [texto]. Devuelve cuántos borró. */
    @Synchronized
    fun olvidar(texto: String): Int {
        val antes = datos.size
        datos.removeAll { it.first.contains(texto, ignoreCase = true) }
        guardar()
        return antes - datos.size
    }

    @Synchronized
    fun comoJson(): JSONArray = JSONArray().apply {
        datos.forEach { (dato, fecha) -> put(JSONObject().put("dato", dato).put("fecha", fecha)) }
    }

    /** Añade los recuerdos de otro dispositivo (la notebook) que aún no tenga. Devuelve cuántos añadió. */
    @Synchronized
    fun fusionar(otros: JSONArray): Int {
        val conocidos = datos.map { it.first.trim().lowercase() }.toMutableSet()
        var anadidos = 0
        for (i in 0 until otros.length()) {
            val otro = otros.optJSONObject(i) ?: continue
            val dato = otro.optString("dato").trim()
            if (dato.isEmpty() || !conocidos.add(dato.lowercase())) continue
            datos += dato to otro.optString("fecha").ifEmpty { LocalDate.now().toString() }
            anadidos++
        }
        if (anadidos > 0) guardar()
        return anadidos
    }

    @Synchronized
    fun comoTexto(): String =
        if (datos.isEmpty()) "(todavía no sabes nada del usuario)"
        else datos.joinToString("\n") { (dato, fecha) -> "- $dato (anotado el $fecha)" }

    private fun guardar() {
        // Mismo formato que datos/memoria.json del Jarvis del PC
        val lista = JSONArray()
        datos.forEach { (dato, fecha) -> lista.put(JSONObject().put("dato", dato).put("fecha", fecha)) }
        val texto = lista.toString(2)
        archivo.writeText(texto)
        copia?.runCatching { writeText(texto) }
    }

    companion object {
        private const val NOMBRE = "memoria.json"
    }
}

File: jarvis/memoria.py
"""Memoria a largo plazo: lo que Jarvis sabe de ti entre una sesión y otra.

Se comparte con la app del celular (servidor.py): los recuerdos de los dos se fusionan.
"""

import json
import threading
from datetime import date
from pathlib import Path

from .config import DIR_DATOS


class Memoria:
    def __init__(self, archivo: Path = DIR_DATOS / "memoria.json"):
        self._archivo = archivo
        self._cerrojo = threading.Lock()  # la usan a la vez la conversación y el servidor
        self._datos: list[dict] = []
        if archivo.exists():
            self._datos = json.loads(archivo.read_text(encoding="utf-8"))

    def recordar(self, dato: str) -> None:
        with self._cerrojo:
            self._datos.append({"dato": dato.strip(), "fecha": date.today().isoformat()})
            self._guardar()

    def olvidar(self, texto: str) -> int:
        """Borra los recuerdos que contienen `texto`. Devuelve cuántos borró."""
        with self._cerrojo:
            antes = len(self._datos)
            self._datos = [d for d in self._datos if texto.lower() not in d["dato"].lower()]
            self._guardar()
            return antes - len(self._datos)

    def datos(self) -> list[dict]:
        with self._cerrojo:
            return [dict(d) for d in self._datos]

    def fusionar(self, otros: list[dict]) -> int:
        """Añade los recuerdos de otro dispositivo que aún no tenga. Devuelve cuántos añadió."""
        with self._cerrojo:
            conocidos = {d["dato"].strip().lower() for d in self._datos}
            nuevos = [{"dato": d["dato"].strip(), "fecha": d.get("fecha", date.today().isoformat())}
                      for d in otros
                      if isinstance(d, dict) and str(d.get("dato", "")).strip()
                      and d["dato"].strip().lower() not in conocidos]
            if nuevos:
                self._datos.extend(nuevos)
                self._guardar()
            return len(nuevos)

    def como_texto(self) -> str:
        with self._cerrojo:
            if not self._datos:
                return "(todavía no sabes nada del usuario)"
            return "\n".join(f"- {d['dato']} (anotado el {d['fecha']})" for d in self._datos)

    def _guardar(self) -> None:
        self._archivo.parent.mkdir(parents=True, exist_ok=True)
        self._archivo.write_text(
            json.dumps(self._datos, ensure_ascii=False, indent=2), encoding="utf-8"
        )
